Colbert local mode support both as retriever and reranker.

Athe-kunal opened the pull request:
Here are the changes proposed here

1. Colbert as a local retriever
2. Colbert as a local reranker
3. Not treating reranking as first-class citizen, rather added RetrieveThenRerank apart from Retrieve for retrieving and reranking
4. Attached a jupyter notebook for showing the implementation details

Athe-kunal commented:
Fixes and Features in the PR

The notebook to follow with all the changes is [here](https://github.com/Athe-kunal/dspy/blob/main/examples/integrations/colbert/colbert_local.ipynb)

1. [Here](https://github.com/stanfordnlp/dspy/blob/main/dsp/primitives/search.py#L69), the retriever for retrieveEnsemble and retrieve, combines the passages from all the queries into one list. It is erroneous, as the user who is passing multiple queries, expects that he/she will get the relevant passages for each query in the queries list. In the PR, I have fixed it to return a prediction object or prediction objects in a list where each prediction object contains relevant passages for the given query
2. Also, the retriever only returns the text without metadata. The metadata will be helpful in downstream tasks of source citations. The metadata return support was added, but the rest of the pipeline won't change. Example of current output

```python
[Prediction(
     pid=[6, 48, 74, 47, 33],
     rerank_score=[15.8359375, 14.2109375, 12.5703125, 11.7890625, 9.1796875],
     passages=['The best things in life are free.', 'Patience is a virtue.', 'To be or not to be, that is the question.', 'Out of sight, out of mind.', 'No pain, no gain.']
 ),
 Prediction(
     pid=[33, 0, 47, 74, 16],
     rerank_score=[19.828125, 12.2890625, 11.171875, 9.09375, 6.8984375],
     passages=['No pain, no gain.', "It's a piece of cake.", 'Out of sight, out of mind.', 'To be or not to be, that is the question.', 'Keep your friends close and your enemies closer.']
 )]
```
3. Currently, the reranker will not work in retrieveEnsemble as it is expecting the query and the passages. But there is no reranker support. In this PR, I have added support for Colbert as a reranker. An example of it below

```python
colbert_config = ColBERTConfig()
colbert_config.index_name = 'colbert-ir-index'
colbert_reranker = dspy.ColBERTv2RerankerLocal(
    checkpoint='colbert-ir/colbertv2.0',colbert_config=colbert_config)

dspy.settings.configure(rm=colbert_retriever,reranker=colbert_reranker)

retrieve_rerank = dspy.RetrieveThenRerank(k=5)

pred = retrieve_rerank(
    ["What is the meaning of life?","Meaning of pain?"]
)
```

The `RetrieveThenRerank` will first retrieve, and then from those passages will rerank using the Max-sim operator in ColBERT. Other reranker integrations will be next. The idea about RetrieveThenRerank was mentioned as a TO-DO [here](https://github.com/stanfordnlp/dspy/blob/696f2d2e1f96f173abb36302c71a725d564dfadb/dsp/utils/settings.py#L27).

@okhat, @arnavsinghvi11 @CShorten. Please review the PR and suggest feedbacks on improving it.

Josephrp reviewed (approved):
very nice work 👏🏻👏🏻

Athe-kunal commented:
> very nice work 👏🏻👏🏻

Thanks @Josephrp

arnavsinghvi11 commented on examples/integrations/colbert/colbert_local.ipynb:
```diff
@@ -0,0 +1,751 @@
+{
```
would be good to remove the outputs from some of the notebook cells if not entirely relevant for someone to understand the tutorial notebook. let's try to keep dspy-specific cell outputs that help users understand the notebook easily.

arnavsinghvi11 replied:
@Athe-kunal

arnavsinghvi11 commented:
Thanks a lot for these additions @Athe-kunal ! Really appreciate the tutorial notebook! I've left some comments for the PR to mainly address some code cleanup changes. 

It would also be great if you could add some documentation for the local ColBERT models to the Retrieval documentation. You can follow the existing [ColBERTv2](https://github.com/stanfordnlp/dspy/blob/197a2c2b57f04f61e1ab03b0d1c871cea68a6b4d/docs/api/retrieval_model_clients/ColBERTv2.md) docs for reference and it would be great to add more specifics on the user parameters for interacting with the RM and potentially some of the relevant implementation details. Thanks!

Athe-kunal commented on dsp/modules/colbertv2.py:
```diff
@@ -74,3 +75,118 @@ def colbertv2_post_request_v2_wrapped(*args, **kwargs):
+            load_only (bool, optional): whether to load the index or . Defaults to False.
+            index_name (str, optional): name of the index. Defaults to "colbert_rm".
+            checkpoint (str, optional): checkpoint for generating embeddings. Defaults to 'colbert-ir/colbertv2.0'.
+            colbert_config (ColBERTConfig, optional): colbert config for building and searching. Defaults to ColBERTConfig().
```
Ideally, it should be typecasted to ColBERTConfig, but for that, I need to import it inside the init function, thus it needs to be installed inside the requirements file, which is not ideal. Hence I default it to none, but I have added an assert to make sure that users are not passing a invalid config

Athe-kunal commented on dsp/modules/colbertv2.py:
```diff
@@ -74,3 +75,118 @@ def colbertv2_post_request_v2_wrapped(*args, **kwargs):
+
+    def __call__(self,query:str,passages:List[str]=[]):
+        import numpy as np
+        from colbert.modeling.colbert import ColBERT
```
I think users may use Colbert as a retriever and as a reranker separately, hence it would be helpful that we check the import errors separately. Please let me know if I am right in my assumption.

Athe-kunal commented on dsp/primitives/search.py:
```diff
@@ -9,17 +10,21 @@ def retrieve(query: str, k: int, **kwargs) -> list[str]:
     if not dsp.settings.rm:
         raise AssertionError("No RM is loaded.")
+    if not dsp.settings.reranker:
+        warnings.warn("If you want to use the Reranker, please use dspy.RetrieveThenRerank")
```
@arnavsinghvi11, can you explain what you mean by logging?
Do I have to create a python logging object file and then log these? Sorry if this is a trivial question

arnavsinghvi11 replied:
https://github.com/stanfordnlp/dspy/blob/d09d984ecaf17f7262294d50fe46fd8105fbf291/dspy/evaluate/evaluate.py#L56 - feel free to reference this

Athe-kunal replied:
The dspy logger object is not available in the dsp folder, hence I followed logging as done [here](https://github.com/stanfordnlp/dspy/blob/6ef5e37d75e3555d91905401bfc374e45ab68282/dsp/modules/anthropic.py#L16) for anthropic LM. Is there a better way to log this?

Athe-kunal commented on dsp/primitives/search.py:
```diff
@@ -9,17 +10,21 @@ def retrieve(query: str, k: int, **kwargs) -> list[str]:
-        passages_cs_scores = dsp.settings.reranker(query, passages)
-        passages_cs_scores_sorted = np.argsort(passages_cs_scores)[::-1]
-        passages = [passages[idx] for idx in passages_cs_scores_sorted]
+    # if dsp.settings.reranker:
```
The functions are a drop-in replacement for the other older functions with support for returning metadata for source citations. So I think we don't need to show a deprecation warning. Please let me know if I am thinking it right

arnavsinghvi11 replied:
hmm, i think it's not useful to keep commented code and have users identify what changes were made. i recommend my earlier suggestion to maintain consistency

Athe-kunal replied:
I have added the commented codes back for future reference.

Athe-kunal commented on dspy/retrieve/retrieve.py:
```diff
@@ -29,14 +29,89 @@ def load_state(self, state):
+        # print(queries)
+        # TODO: Consider removing any quote-like markers that surround the query too.
+        k = k if k is not None else self.k
+        passages = dsp.retrieveRerankEnsemble(queries, k=k,**kwargs)
```
Sorry I was not able to understand this. @arnavsinghvi11 
Do you want to have a common utility function for both `Retrieve` and `RetrieveTheRerank` to process the returned documents?

arnavsinghvi11 replied:
Hi @Athe-kunal , yeah it seemed like there is some repetitive code in both forward passes that can be abstracted out for the different retriever types. let me know if this change makes sense

Athe-kunal replied:
Hi @arnavsinghvi11 
I have abstracted the repetitive code part. However there are some nuances in the multi-query retriever, hence I didn't make a helper function for it. But for a single query, I have added a helper function `single_query_passage`. Please let me know if I need to make some other changes.

Athe-kunal commented:
Thanks a ton, @arnavsinghvi11 for this detailed feedback on my PR. It certainly helps me to be a better contributor. I have made the changes and also addressed some confusion, please o help me out there. I will work on the documentation changes for Colbert. Looking forward to more collaboration

Athe-kunal commented:
@arnavsinghvi11 
I have added the documentation for COlbert. Please do review this and suggest edits

Athe-kunal commented:
@arnavsinghvi11 Can you please review it?

Athe-kunal commented:
Hi @arnavsinghvi11 
It has been a while, can you review the changes here?

arnavsinghvi11 commented on dsp/primitives/search.py:
```diff
@@ -28,44 +33,67 @@ def retrieveRerankEnsemble(queries: list[str], k: int,**kwargs) -> list[str]:
         raise AssertionError("No RM is loaded.")
-    if dsp.settings.reranker:
-        return retrieveRerankEnsemble(queries, k)
+    if not dsp.settings.reranker:
```
as with above - https://github.com/stanfordnlp/dspy/blob/d09d984ecaf17f7262294d50fe46fd8105fbf291/dspy/evaluate/evaluate.py#L56

Athe-kunal replied:
as with above-
The dspy logger object is not available in the dsp folder, hence I followed logging as done [here](https://github.com/stanfordnlp/dspy/blob/6ef5e37d75e3555d91905401bfc374e45ab68282/dsp/modules/anthropic.py#L16) for anthropic LM. Is there a better way to log this?

arnavsinghvi11 commented:
Hi @Athe-kunal , just left some follow up comments. it seems that there are still some leftover comments to address. feel free to discuss them more directly here as needed.

Athe-kunal commented:
Hi @arnavsinghvi11 
I have made the required changes and resolved some issues. Please let me know if this is good to merge.

ahmed-moubtahij commented:
Great PR! Hope it gets merged soon, I find myself needing this.

Athe-kunal commented:
@arnavsinghvi11 
Can you review this again?

arnavsinghvi11 commented:
Hi @Athe-kunal , thanks again. I'd love to merge this PR but unfortunately, it breaks some existing caches we have in the repository, particularly the intro.ipynb. For more clarity, the changes to [search.py](https://github.com/stanfordnlp/dspy/pull/797/files#diff-937571e5fdca9032d4f6962c738bbc643cbc6043a2e623b73760d415b850ffdb) and [retrieve.py](https://github.com/stanfordnlp/dspy/pull/797/files#diff-6ba09b368b7910a57fa1cd0ce8b6112900ac4424a1823d2b68ab440517b9f88a) break the cached retrieved outputs in the intro notebook (which we keep fully cached so users can interact with the DSPy tutorial without having to expend their API key credits). 

If you could make changes to this PR that enable the existing search/retrieve functions as I mentioned earlier, while introducing the new behavior separately, I can merge it after that. lmk if this makes sense!

Athe-kunal commented:
Thanks for the suggestion @arnavsinghvi11 
I had a query regarding the joblib memory cache. In the current implementation, we have a list of text (only the relevant context) which is being cached. But in my implementation, I am returning a prediction object. Can joblib cache any abstracted data type or just text?
Also, for the local Colbertv2. Should I integrate the caching mechanism? It is a local model, and it does not require us to send the request to an API. Can you help me on the PR @arnavsinghvi11?

Athe-kunal commented:
Hi @arnavsinghvi11 
Can you help me with the caching functionality here for the PR? I am unable to understand the caching mechanism here.

arnavsinghvi11 commented:
Hi @Athe-kunal , 

I would firstly recommend keeping the `search` and `retrieve` abstractions with as minimal differences as possible. I think I mentioned this in an earlier review but the changes to `retrieve`, `retrieveRerankEnsemble`, `retrieveEnsemble`, and the `forward` function in `Retrieve` directly impact existing caches with the `intro.ipynb` notebook. Any changes to those make it difficult to merge PRs without having to rerun all the requests with the new changes (which we highly prefer not to do to maintain consistency). 

In more simple terms, the `ColBERTv2` pipeline with `search` and `retrieve` should work as-is, and you can test this with your PR changes by running the intro notebook and checking if any changes added result in the notebook failing to execute. 
Let me know if this is clear as it's not a _caching_ problem to implement in this PR, but rather to maintain existing behavior that will ensure existing caches work!

Some steps I suggest with this is reverting back to the original behavior of search and retrieve and then integrating `ColBERTv2RetrieverLocal` and `ColBERTv2RerankerLocal` to work with those untouched functions. If you follow the existing setup of `ColBERTv2` with its corresponding return types, I think this can work. Once that's done, feel free to ping me back on the PR so we can take a look and ensure no existing caches break. 

From there, the next step would be to mirror the existing colbertv2 cache functions to align with the local mode being introduced. 

Let me know if this all makes sense!

Athe-kunal commented:
Hi @arnavsinghvi11 
Thanks for your detailed description. I had one small query; the Colbert retriever is a local model. The caching is essential if we are sending to a third party API. But if it is a local model, then does it require caching? If users expose the colbert model to a server, then they can directly use previous ColbertRetriever which has caching. Please let me know if I am
missing something.

arnavsinghvi11 commented:
>But if it is a local model, then does it require caching?

Not a direct answer, but [HFClientTGI](https://github.com/stanfordnlp/dspy/blob/0e595a70cd9d14b0db25252320aad11eae08fd72/dsp/modules/hf_client.py#L105) supports caching for models hosted locally (REST API). Maybe it would benefit here as well. But caching is not a requirement with this PR. Ensuring backward compatibility with the existing `dspy.ColBERTv2` in addition to supporting local ColBERT is!

Athe-kunal commented:
Got it @arnavsinghvi11, working on it
Also, if you don't mind, can you share your discord handle on the DSPy discord channel, it would be helpful to interact there too.

Athe-kunal commented:
Hi @arnavsinghvi11 
I tested with the `intro.ipynb`  file, and it worked without hiccups.
Can you respond to the above requests?

arnavsinghvi11 commented:
Hi @Athe-kunal , I don't see any changes to the PR. could you check if they are pushed? my username is bigman11 btw

Athe-kunal commented:
Hi @arnavsinghvi11 
I have made the following changes

1. Added separate functions for retrieving with metadata so that existing cache won't break. I am passing `with_metadata` parameter, which has a default value of `False`, thus the current tutorials would work fine with it
2. Also, I added a `by_prob` parameter with default value of `True` to `dspy.Retrieve`. It turns out that `dsp.retrieve` needs this by_prob parameter, but it is being passed as a kwargs parameter.

This PR has become way too much intractable, I will work on ColbertLocal caching in another PR. Can you review this one @arnavsinghvi11 ?

Athe-kunal commented:
Hi @arnavsinghvi11 
Thanks for your support and guidance throughout the PR process, it was a great learning experience. You can squash and merge the changes for this PR.
I will work on the caching for the new Colbert Local models this weekend, but for now you can merge this one

Thanks again for helping me in this PR.

arnavsinghvi11 commented:
Merged. Thanks again @Athe-kunal !
